WIP: Add MVP structure and retrofitting TFL API service

akwc10 opened the pull request.

akwc10 committed:
Added MVP structure and retrofitting TFL journey planner API

akwc10 committed:
MVP structure for MainActivity and retrofitting API service

akwc10 committed:
Moved Constants

fsbarata commented on app/src/main/java/com/my/journeyplanner/Constants.kt:
```diff
@@ -0,0 +1,5 @@
+package com.my.journeyplanner
+
+const val TFL_API_BASE_URL = "https://api.tfl.gov.uk/"
```
probably better to put this in the Api interface

akwc10 replied:
Moved back to interface JourneyPlannerApiService

fsbarata commented on app/src/main/java/com/my/journeyplanner/helpers/JourneyPlannerApiService.kt:
```diff
@@ -0,0 +1,27 @@
+        @Path("fromLocation") fromLocation: String,
+        @Path("toLocation") toLocation: String,
+        @Query("app_id") appId: String,
+        @Query("app_key") appKey: String
```
if these 2 parameters are required in all requests of the api, it will be easier to write an interceptor that adds them

akwc10 replied:
Created:
class ParametersInterceptor : Interceptor {
    override fun intercept(chain: Interceptor.Chain): Response {
        val url = chain.request().url().newBuilder()
            .addQueryParameter("app_id", TFL_API_APP_ID)
            .addQueryParameter("app_key", TFL_API_APP_KEY)
            .build()
        val request = chain.request().newBuilder().url(url).build()

        return chain.proceed(request)
    }
}

fsbarata commented on app/src/main/java/com/my/journeyplanner/helpers/JourneyPlannerApiService.kt:
```diff
@@ -0,0 +1,27 @@
+    ): Call<JourneyPlannerDisambiguationResult>
+
+    companion object {
+        fun apiService(): JourneyPlannerApiService = Retrofit.Builder()
```
any reason why it wouldn't be a constant?

Also since its a factory method I recommend you prefix with create

akwc10 replied:
Still thinking the java way. Updated to:
val createApiService: JourneyPlannerApiService = Retrofit.Builder()...

fsbarata commented on app/src/main/java/com/my/journeyplanner/models/JourneyPlannerDisambiguationResult.kt:
```diff
@@ -0,0 +1,135 @@
+import com.squareup.moshi.Json
+
+data class JourneyPlannerDisambiguationResult(
+    @Json(name = "fromLocationDisambiguation")
```
you dont need this since it matches the name of the field

akwc10 replied:
Removed all @Json() annotations

fsbarata commented on app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt:
```diff
@@ -0,0 +1,25 @@
+    }
+
+    override fun onPlanMyJourneyClicked() {
+        JourneyPlannerResultsAsyncTask().execute(
```
how do you cancel this?

akwc10 replied:
Updated to use retrofit async task that provides .cancel()

fsbarata commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -32,7 +32,7 @@
         app:layout_constraintEnd_toEndOf="parent"
         app:layout_constraintStart_toStartOf="parent"
-        app:layout_constraintTop_toBottomOf="@+id/editTextFrom" />
+        app:layout_constraintTop_toBottomOf="@+id/editTextFromLocation" />
```
do not use @+id in references, use @id. + will create when it doesnt exist which leads to bugs.

akwc10 replied:
Updated as advised

akwc10 committed:
Updated apiService, added Interceptor, cleaned model, removed @+id references

akwc10 committed:
Retrofit async call and cancelling

fsbarata commented on app/build.gradle:
```diff
@@ -24,11 +24,13 @@ android {
     testImplementation 'junit:junit:4.12'
     androidTestImplementation 'androidx.test:runner:1.2.0'
     androidTestImplementation 'androidx.test.espresso:espresso-core:3.2.0'
+    implementation 'com.squareup.retrofit2:retrofit:2.4.0'
```
this version should be extracted like you did for the kotlin version since it's common to all retrofit libraries

akwc10 replied:
Extracted to top level build.gradle
ext.retrofit_version = '2.4.0'

fsbarata commented on app/src/main/java/com/my/journeyplanner/MainActivity.kt:
```diff
@@ -1,12 +1,68 @@
 
-class MainActivity : AppCompatActivity() {
+val apiService by lazy { JourneyPlannerApiService.createApiService }
+var call: Call<JourneyPlannerDisambiguationResult>? = null
```
oh boy, this is a terrible way to track state :) You should never have global vars. No exceptions

akwc10 replied:
I was following an example. Wondered why it didn't feel right. Moved back to presenter as advised in later comment

fsbarata commented on app/src/main/java/com/my/journeyplanner/MainActivity.kt:
```diff
@@ -1,12 +1,68 @@
+import retrofit2.Call
 
-class MainActivity : AppCompatActivity() {
+val apiService by lazy { JourneyPlannerApiService.createApiService }
```
what is the reason why this is lazy. Why did you put this here

akwc10 replied:
Whoops!. Shouldn't be lazy as it's needed all the time. Moved back to presenter as advised in  later comment

fsbarata commented on app/src/main/java/com/my/journeyplanner/MainActivity.kt:
```diff
@@ -1,12 +1,68 @@
+
+    override fun onDestroy() {
+        super.onDestroy()
+        call?.cancel()
```
this should call the presenter. Cancellation is done onStop, not destroy

akwc10 replied:
In Mainactivity:
    override fun onStop() {
        super.onStop()
        mainPresenter.cancelAsyncCall()
    }

akwc10 replied:
I see, so it triggers when you background app

fsbarata commented on app/src/main/java/com/my/journeyplanner/MainActivity.kt:
```diff
@@ -1,12 +1,68 @@
+            if (actionId == EditorInfo.IME_ACTION_DONE ||
+                (event.keyCode == KeyEvent.KEYCODE_ENTER && event.action == KeyEvent.ACTION_DOWN)
+            ) {
+                call = mainPresenter.onPlanMyJourneyClicked()
```
call should be tracked in the presenter

akwc10 replied:
Moved:
private var call: Call<JourneyPlannerDisambiguationResult>? = null
to MainPresenter

fsbarata commented on app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt:
```diff
@@ -0,0 +1,53 @@
+        call.enqueue(object : Callback<JourneyPlannerDisambiguationResult> {
+            override fun onFailure(call: Call<JourneyPlannerDisambiguationResult>, throwable: Throwable) {
+                Log.e(TAG, "Other error: ${throwable.message}")
+                Log.e(TAG, "Other error: ${throwable.stackTrace.forEach { println(it) }}")
```
mark TODOs with the intended future changes

akwc10 replied:
Added:
/*            TODO("Default adapter treats response code 300 Multiple Options as unsuccessful and")
            TODO("puts response in the errorBody but is actually a valid response.")
            TODO("Implement custom adapter to parse the valid response in the errorBody")   */

fsbarata commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -32,16 +32,16 @@
-        android:textSize="18sp"
         android:text="@string/textView_leaving_now"
-        app:layout_constraintBaseline_toBaselineOf="@+id/buttonChangeTime"
+        android:textSize="18sp"
```
use material text appearances

akwc10 replied:
Replaced with:
android:textAppearance="@style/TextAppearance.AppCompat.Body1"

That's really handy

akwc10 committed:
Async call state moved to presenter, call cancelled in onStop(), minor changes

Async call state was global variable i.e. very bad.
Call was cancelled in onDestroy() which is incorrect.
Use TextAppearance instead of specifying custom font size.
Dedupe gradle using retrofit_version constant.

akwc10 committed:
Forgot to include MainPresenter in previous commit

fsbarata commented on app/src/main/java/com/my/journeyplanner/helpers/JourneyPlannerApiService.kt:
```diff
@@ -0,0 +1,28 @@
+        private const val TFL_API_BASE_URL = "https://api.tfl.gov.uk/"
+        private val client = OkHttpClient().newBuilder().addInterceptor(ParametersInterceptor()).build()
+
+        val createApiService: JourneyPlannerApiService = Retrofit.Builder()
```
fun createApiService()

akwc10 replied:
Reverted to java style function

akwc10 committed:
Revert createApiService to function, use MaterialComponents text style

akwc10 committed:
MaterialComponents dependency
